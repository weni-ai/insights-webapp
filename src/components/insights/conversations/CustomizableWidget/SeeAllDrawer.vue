<template>
  <UnnnicDrawer
    class="see-all-drawer"
    data-testid="see-all-drawer"
    :modelValue="modelValue"
    title="CUSTOM"
    size="lg"
    @close="emit('update:modelValue', false)"
  >
    <template #content>
      <ProgressTable
        :progressItems="formattedData"
        data-testid="see-all-drawer-table"
      />
    </template>
  </UnnnicDrawer>
</template>

<script setup lang="ts">
import ProgressTable from '@/components/ProgressTable.vue';
import { useWidgetFormatting } from '@/composables/useWidgetFormatting';
import type { CsatResult } from '@/services/api/resources/conversational/widgets';

const { formatPercentage, formatNumber } = useWidgetFormatting();

const props = defineProps<{
  modelValue: boolean;
  data: CsatResult[];
}>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const defaultColor = '#3182CE';
const defaultBackgroundColor = '#BEE3F8';

const formattedData = props.data.map((item) => ({
  label: item.label,
  value: item.value,
  description: `${formatPercentage(item.value)} (${formatNumber(item.full_value)})`,
  backgroundColor: defaultBackgroundColor,
  color: defaultColor,
  subItems: [],
}));
</script>
