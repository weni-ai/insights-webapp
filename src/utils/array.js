export function sortByKey(arr, key, order = 'asc') {
  const compareString = arr.every((item) => typeof item[key] === 'string');
  const ordenedArr = compareString
    ? arr.sort((a, b) => a[key].localeCompare(b[key]))
    : arr.toSorted((a, b) => a[key] - b[key]);
  return order === 'desc' ? ordenedArr.reverse() : ordenedArr;
}
