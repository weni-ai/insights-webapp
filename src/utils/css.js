export function pxToVh(px) {
  const vh = window.innerHeight;
  const pxAsVh = (px / vh) * 100;
  return pxAsVh;
}
