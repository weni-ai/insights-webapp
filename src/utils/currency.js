export const currencySymbols = {
  BRL: 'R$',
  ARS: '$',
  USD: '$',
  EUR: 'â‚¬',
};
