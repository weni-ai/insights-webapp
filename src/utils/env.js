export default function env(name) {
  if (import.meta.env[name] && name === 'VITE_FIREBASE_CONFIG')
    return JSON.parse(import.meta.env[name]);

  return (
    window?.configs?.[name] ||
    import.meta.env[name] ||
    window?.configs?.[`VITE_${name}`] ||
    import.meta.env[`VITE_${name}`]
  );
}
