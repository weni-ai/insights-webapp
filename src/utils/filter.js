export function isFilteringDates({ appliedFilters, currentDashboardFilters }) {
  const currentDashboardDateFiltersNames = currentDashboardFilters
    .filter((filter) => filter.type === 'date_range')
    .map((filter) => filter.name);

  const appliedFilterNames = Object.keys(appliedFilters);

  return !!currentDashboardDateFiltersNames.some(
    (currentDashboardDateFilterName) =>
      appliedFilterNames.some((appliedFilterName) =>
        currentDashboardDateFilterName.includes(appliedFilterName),
      ),
  );
}
