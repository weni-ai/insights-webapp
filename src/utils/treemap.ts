import i18n from '@/utils/plugins/i18n';
import type { topicDistributionMetric } from '@/services/api/resources/conversational/topics';

/**
 * Sorts, limits to 5 items, and groups the rest under "Others".
 * @param data Original array
 * @returns New array ready to use
 */
export function prepareTopData(
  data: topicDistributionMetric[],
): topicDistributionMetric[] {
  const sorted = [...data].sort((a, b) => b.value - a.value);

  if (sorted.length <= 5) {
    return sorted;
  }

  const topFive = sorted.slice(0, 5);
  const rest = sorted.slice(5);

  const othersValue = rest.reduce((sum, item) => sum + item.value, 0);
  const othersPercentage = rest.reduce(
    (sum, item) => sum + (item.percentage ?? 0),
    0,
  );

  topFive.push({
    label: i18n.global.t('conversations_dashboard.others'),
    value: othersValue,
    percentage: othersPercentage,
    uuid: '',
  });
  return topFive;
}

interface DataWithColor extends topicDistributionMetric {
  color: string;
  hoverColor: string;
}

const COLOR_PALETTE = {
  position: [
    '#C6FFF7', // color-weni-100
    '#E9D8FD', // color-aux-purple-100
    '#BEE3F8', // color-aux-blue-100
    '#C6F6D5', // color-aux-green-100
    '#FEEBC8', // color-aux-orange-100
  ],
  positionHover: [
    '#4DFBEA', // color-weni-300
    '#B794F4', // color-aux-purple-300
    '#63B3ED', // color-aux-blue-300
    '#68D391', // color-aux-green-300
    '#F6AD55', // color-aux-orange-300
  ],
  special: {
    others: {
      normal: '#E2E6ED', // color-neutral-soft
      hover: '#D0D3D9', // color-neutral-cleanest
    },
    unclassified: {
      normal: '#FED7D7', // color-aux-red-100
      hover: '#FC8181', // color-aux-red-300
    },
  },
  fallback: {
    normal: '#E2E6ED', // color-neutral-soft
    hover: '#D0D3D9', // color-neutral-cleanest
  },
};

/**
 * Determines if a label is a special case (others or unclassified)
 */
function isSpecialLabel(label: string): 'others' | 'unclassified' | null {
  if (label.toLowerCase() === 'other') {
    return 'others';
  }
  if (
    label.toLowerCase() ===
    i18n.global.t('conversations_dashboard.unclassified')
  ) {
    return 'unclassified';
  }
  return null;
}

/**
 * Gets colors for an item based on its label and position
 */
function getItemColors(
  label: string,
  index: number,
): { color: string; hoverColor: string } {
  const specialType = isSpecialLabel(label);

  if (specialType) {
    const specialColors = COLOR_PALETTE.special[specialType];
    return {
      color: specialColors.normal,
      hoverColor: specialColors.hover,
    };
  }

  return {
    color: COLOR_PALETTE.position[index] ?? COLOR_PALETTE.fallback.normal,
    hoverColor:
      COLOR_PALETTE.positionHover[index] ?? COLOR_PALETTE.fallback.hover,
  };
}

/**
 * Applies color based on label position or rules
 * @param data Sorted array
 * @returns New array with color field
 */
export function addColors(data: topicDistributionMetric[]): DataWithColor[] {
  return data.map((item, index) => ({
    ...item,
    label:
      item?.label?.toLowerCase() === 'other'
        ? i18n.global.t('conversations_dashboard.others')
        : item.label,
    ...getItemColors(item.label, index),
  }));
}
