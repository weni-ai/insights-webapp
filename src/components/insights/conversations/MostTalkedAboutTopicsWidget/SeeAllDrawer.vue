<template>
  <UnnnicDrawer
    class="see-all-drawer"
    :modelValue="modelValue"
    :title="$t('conversations_dashboard.most_talked_about_topics.title')"
    size="lg"
    @close="emit('update:modelValue', false)"
  >
    <template #content>
      <ProgressTable
        :progressItems="MOCK_DATA_PROGRESS_TABLE"
        :subItemsDescription="
          $t('conversations_dashboard.most_talked_about_topics.subtopics')
        "
        :expandedItems="expandedItems"
      />
    </template>
  </UnnnicDrawer>
</template>

<script setup lang="ts">
import ProgressTable from '@/components/ProgressTable.vue';
import { addColors } from '@/utils/treemap';
import { useWidgetFormatting } from '@/composables/useWidgetFormatting';

const { formatPercentage, formatNumber } = useWidgetFormatting();

defineProps<{
  modelValue: boolean;
}>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const MOCK_DATA = [
  {
    label: 'Entrega atrasada',
    value: 6973,
    percentage: 29,
    subItems: [
      {
        label: 'Atraso no transporte',
        value: 5229,
        percentage: 75,
      },
      {
        label: 'Problema na preparação',
        value: 1000,
        percentage: 25,
      },
    ],
  },
  {
    label: 'Produto defeituoso',
    value: 5500,
    percentage: 23,
  },
  {
    label: 'Dúvidas sobre preço',
    value: 1600,
    percentage: 16,
    subItems: [
      {
        label: 'Desconto no valor',
        value: 1825,
        percentage: 35,
      },
      {
        label: 'Desconto no frete',
        value: 375,
        percentage: 65,
      },
    ],
  },
  {
    label: 'Cancelamento',
    value: 1400,
    percentage: 14,
  },
  {
    label: 'Unclassified',
    value: 1000,
    percentage: 13,
  },
  {
    label: 'Outros',
    value: 500,
    percentage: 5,
  },
];

const coloredData = addColors(MOCK_DATA);

const MOCK_DATA_PROGRESS_TABLE = coloredData.map((item) => ({
  label: item.label,
  value: item.percentage,
  description: `${formatPercentage(item.percentage)} (${formatNumber(item.value)})`,
  backgroundColor: item.color,
  color: item.hoverColor,
  subItems: MOCK_DATA.find(
    (mockItem) => mockItem.label === item.label,
  )?.subItems?.map((subItem) => ({
    label: subItem.label,
    value: subItem.percentage,
    description: `${formatPercentage(subItem.percentage)} (${formatNumber(subItem.value)})`,
    backgroundColor: item.color,
    color: item.hoverColor,
  })),
}));

const expandedItems = defineModel<string[]>('expandedItems', {
  default: () => [],
});
</script>
