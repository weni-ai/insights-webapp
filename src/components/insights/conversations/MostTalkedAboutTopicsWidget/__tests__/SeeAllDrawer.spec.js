import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import SeeAllDrawer from '../SeeAllDrawer.vue';

const mockWidgetFormatting = {
  formatPercentage: vi.fn((value) => `${value}%`),
  formatNumber: vi.fn((value) => value.toString()),
};

vi.mock('@/composables/useWidgetFormatting', () => ({
  useWidgetFormatting: () => mockWidgetFormatting,
}));

vi.mock('@/utils/treemap', () => ({
  addColors: vi.fn((data) =>
    data.map((item, index) => ({
      ...item,
      color: `#color${index}`,
      hoverColor: `#hover${index}`,
    })),
  ),
}));

import { addColors } from '@/utils/treemap';
const mockAddColors = addColors;

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      conversations_dashboard: {
        most_talked_about_topics: {
          title: 'Most Talked About Topics',
          subtopics: 'Subtopics',
        },
      },
    },
  },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('SeeAllDrawer', () => {
  let wrapper;

  const defaultProps = {
    modelValue: true,
    data: [
      {
        label: 'Technology',
        percentage: 45.2,
        value: 1520,
        subtopics: [
          { label: 'AI', percentage: 25.1, value: 845 },
          { label: 'Web Development', percentage: 20.1, value: 675 },
        ],
      },
      {
        label: 'Business',
        percentage: 32.8,
        value: 1104,
        subtopics: [{ label: 'Marketing', percentage: 15.5, value: 522 }],
      },
      { label: 'Health', percentage: 22.0, value: 740, subtopics: [] },
    ],
    expandedItems: ['Technology'],
  };

  const createWrapper = (props = {}) =>
    mount(SeeAllDrawer, {
      props: { ...defaultProps, ...props },
      global: {
        stubs: {
          UnnnicDrawer: {
            template:
              '<div data-testid="topics-see-all-drawer" class="see-all-drawer" @click="$emit(\'close\')"><slot name="content" /></div>',
            emits: ['close'],
          },
          ProgressTable: true,
        },
      },
    });

  const drawer = () => wrapper.find('[data-testid="topics-see-all-drawer"]');
  const progressTable = () =>
    wrapper.find('[data-testid="topics-see-all-table"]');

  beforeEach(() => {
    vi.clearAllMocks();
    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render all elements', () => {
      expect(drawer().exists()).toBe(true);
      expect(progressTable().exists()).toBe(true);
    });

    it('should render with different data sets', () => {
      const customData = [
        {
          label: 'Sports',
          percentage: 60.0,
          value: 2000,
          subtopics: [{ label: 'Football', percentage: 35.0, value: 1200 }],
        },
      ];
      wrapper = createWrapper({ data: customData });

      expect(wrapper.props('data')).toEqual(customData);
      expect(progressTable().exists()).toBe(true);
    });

    it('should render with empty data', () => {
      wrapper = createWrapper({ data: [] });

      expect(wrapper.props('data')).toEqual([]);
      expect(progressTable().exists()).toBe(true);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(drawer().classes()).toContain('see-all-drawer');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Event handling', () => {
    it('should emit update:modelValue when closed', () => {
      wrapper.vm.$emit('update:modelValue', false);

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });

    it('should emit update:modelValue false when drawer close event is triggered', async () => {
      // Simulate clicking the drawer to trigger close event
      await drawer().trigger('click');

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });

    it('should handle expandedItems model updates', () => {
      const newExpandedItems = ['Business', 'Health'];
      wrapper.vm.$emit('update:expandedItems', newExpandedItems);

      expect(wrapper.emitted('update:expandedItems')).toBeTruthy();
      expect(wrapper.emitted('update:expandedItems')[0]).toEqual([
        newExpandedItems,
      ]);
    });
  });

  describe('Data processing', () => {
    it('should call addColors with data', () => {
      expect(mockAddColors).toHaveBeenCalledWith(defaultProps.data);
    });

    it('should format data correctly', () => {
      const expectedFormattedData = [
        {
          label: 'Technology',
          value: 45.2,
          description: '45.2% (1520)',
          backgroundColor: '#color0',
          color: '#hover0',
          subItems: [
            {
              label: 'AI',
              value: 25.1,
              description: '25.1% (845)',
              backgroundColor: '#color0',
              color: '#hover0',
            },
            {
              label: 'Web Development',
              value: 20.1,
              description: '20.1% (675)',
              backgroundColor: '#color0',
              color: '#hover0',
            },
          ],
        },
        {
          label: 'Business',
          value: 32.8,
          description: '32.8% (1104)',
          backgroundColor: '#color1',
          color: '#hover1',
          subItems: [
            {
              label: 'Marketing',
              value: 15.5,
              description: '15.5% (522)',
              backgroundColor: '#color1',
              color: '#hover1',
            },
          ],
        },
        {
          label: 'Health',
          value: 22.0,
          description: '22% (740)',
          backgroundColor: '#color2',
          color: '#hover2',
          subItems: [],
        },
      ];

      expect(wrapper.vm.formattedData).toEqual(expectedFormattedData);
    });

    it('should call formatting functions correctly', () => {
      wrapper.vm.formattedData;

      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(45.2);
      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(32.8);
      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(22.0);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(1520);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(1104);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(740);
    });

    it('should handle subtopics formatting', () => {
      wrapper.vm.formattedData;

      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(25.1);
      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(20.1);
      expect(mockWidgetFormatting.formatPercentage).toHaveBeenCalledWith(15.5);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(845);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(675);
      expect(mockWidgetFormatting.formatNumber).toHaveBeenCalledWith(522);
    });

    it('should pass correct props to ProgressTable', () => {
      expect(progressTable().attributes('progressitems')).toBeDefined();
      expect(progressTable().attributes('expandeditems')).toBeDefined();
      expect(progressTable().attributes('subitemsdescription')).toBeDefined();
    });
  });

  describe('Props validation', () => {
    const propTestCases = [
      {
        modelValue: true,
        data: defaultProps.data,
        expandedItems: ['Technology'],
      },
      { modelValue: false, data: [], expandedItems: [] },
      {
        modelValue: true,
        data: [{ label: 'Single', percentage: 100, value: 500, subtopics: [] }],
        expandedItems: [],
      },
    ];

    propTestCases.forEach(({ modelValue, data, expandedItems }) => {
      it(`should handle props: modelValue=${modelValue} with ${data.length} items and ${expandedItems.length} expanded`, () => {
        wrapper = createWrapper({ modelValue, data, expandedItems });

        expect(wrapper.props('modelValue')).toBe(modelValue);
        expect(wrapper.props('data')).toEqual(data);
        expect(wrapper.props('expandedItems')).toEqual(expandedItems);
      });
    });
  });

  describe('Edge cases', () => {
    it('should handle topics without subtopics', () => {
      const dataWithoutSubtopics = [
        { label: 'Simple Topic', percentage: 50, value: 100, subtopics: [] },
      ];
      wrapper = createWrapper({ data: dataWithoutSubtopics });

      const formattedData = wrapper.vm.formattedData;
      expect(formattedData[0].subItems).toEqual([]);
    });

    it('should handle topics with undefined subtopics', () => {
      const dataWithUndefinedSubtopics = [
        { label: 'Topic', percentage: 50, value: 100 },
      ];
      wrapper = createWrapper({ data: dataWithUndefinedSubtopics });

      const formattedData = wrapper.vm.formattedData;
      expect(formattedData[0].subItems).toBeUndefined();
    });
  });
});
