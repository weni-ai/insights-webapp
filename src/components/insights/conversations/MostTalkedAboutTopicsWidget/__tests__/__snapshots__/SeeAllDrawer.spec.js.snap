// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SeeAllDrawer > Component structure > should match snapshot 1`] = `
<div
  class="see-all-drawer"
  data-testid="topics-see-all-drawer"
  modelvalue="true"
  size="lg"
  title="conversations_dashboard.most_talked_about_topics.title"
>
  
  <progress-table-stub
    data-testid="topics-see-all-table"
    expandeditems="Technology"
    isloading="false"
    progressitems="[object Object],[object Object],[object Object]"
    subitemsdescription="conversations_dashboard.most_talked_about_topics.subtopics"
  />
  
</div>
`;
